kedro-org/kedro-viz
# Refactor CLI with lazy subcommands and deferring imports

## Description

Related to https://github.com/kedro-org/kedro/pull/3883 and resolves a follow up action mentioned [here](https://github.com/kedro-org/kedro/issues/1476#issuecomment-2227978718)

Thanks to @ankatiyar for suggesting the changes and making the POC

## Development notes

- Removed single `package/kedro_viz/launchers/cli.py` file which handles all Kedro-Viz CLI commands 
- Created multiple files to handle each Kedro-Viz CLI command (like package/kedro_viz/launchers/cli/build.py, run.py etc)
- Deferred dependencies until they are required
- Created LazyDefaultGroup class as mentioned [here](https://click.palletsprojects.com/en/8.1.x/complex/#lazily-loading-subcommands)
- Updated tests

> [!Tip]
> Read on the docs for [lazy-loading-subcommands](https://click.palletsprojects.com/en/8.1.x/complex/#lazily-loading-subcommands) and learn more about the initial proposal [here](https://github.com/kedro-org/kedro/issues/1476#issuecomment-2139515747)


## QA notes

- All tests should pass
- Test Kedro-Viz CLI commands (`kedro viz -h`, `kedro viz`, `kedro viz run`, `kedro viz build`, `kedro viz deploy`) with and without options
- Make observations similar to the ones made below -

The lazy subcommand loading and refactoring of viz CLI helped in `kedro` init time as viz cli had some expensive imports all in one file `cli.py`. 

Few observations on demo-project with my system running on `macOS Sonoma (2.4 GHz 8-Core Intel i9, 64GB)` -

**( Comparing after system warmup (initial run) as I am not sure how to clear cache or make this a cold start in the CLI. Please let me know if there is a way to make a cold start again )**

**Before -**

Time taken for `kedro -h` - **4sec**

<img width="1323" alt="image" src="https://github.com/user-attachments/assets/7635cf28-e349-4d5d-a1b9-8418a6a74947">


Time taken for `kedro registry list` - **8sec**

<img width="1324" alt="image" src="https://github.com/user-attachments/assets/3fb8cf93-1da0-4a33-82de-89438d870deb">

Time taken for `kedro viz run` to reach the actual `run` method - **6sec**

To run this, you need [pyinstrument](https://github.com/joerick/pyinstrument) and [speedscope](https://github.com/jlfwong/speedscope)

```bash
# Commands used for profiling

pyinstrument -r speedscope -o profile.speedscope.json -m kedro viz run
speedscope profile.speedscope.json
```

<img width="1790" alt="image" src="https://github.com/user-attachments/assets/1247d972-18dd-488c-8a1c-0ae0b468bcaa">


---------------------

**After -**

Time taken for `kedro -h` - **2sec**

<img width="1335" alt="image" src="https://github.com/user-attachments/assets/d2a478c9-109e-4ebf-a261-45e127e5df6d">

Time taken for `kedro registry list` - **6sec**

<img width="1348" alt="image" src="https://github.com/user-attachments/assets/ed127d1f-9cf0-44d4-8f77-326a523c0a98">

Time taken for `kedro viz run` to reach the actual `run` method - **2.5sec**

To run this, you need [pyinstrument](https://github.com/joerick/pyinstrument) and [speedscope](https://github.com/jlfwong/speedscope)

```bash
# Commands used for profiling

pyinstrument -r speedscope -o profile.speedscope.json -m kedro viz run
speedscope profile.speedscope.json
```

<img width="897" alt="image" src="https://github.com/user-attachments/assets/3d983869-ce09-4a8b-b485-ef058936ccc6">

## Checklist
